data-pipeline/src/pipeline/database.py

```diff
@@ -29,15 +33,149 @@
     query={"driver": "ODBC Driver 18 for SQL Server"} | args,
 )
 
-engine = create_engine(connection_url)
+engine = create_engine(connection_url, fast_executemany=True)
 
 
-@event.listens_for(engine, "before_cursor_execute")
-def receive_before_cursor_execute(
-    conn, cursor, statement, params, context, executemany
-):
-    if executemany:
-        cursor.fast_executemany = True
+@dataclass
+class TempTable:
+    name: str
+    target: str
+    columns: list[str]
+
+    def __str__(self):
+        return self.name
+
+
+def _get_table_columns(table: str) -> list[str]:
+    """
+    Retrieve the columns for a given table.
+
+    The source-of-truth for the table formats should remain the DB
+    itself: here, we retrieve the list of expected columns directly
+    from the table's schema.
+
+    :param table: DB table for which to retrieve columns
+    :return: column names
+    """
+    sql = textwrap.dedent("""
+    SELECT COLUMN_NAME
+      FROM INFORMATION_SCHEMA.COLUMNS
+     WHERE TABLE_NAME = :table
+    ;
+    """).strip()
+
+    with engine.begin() as cnx:
+        results = (
+            cnx.execute(
+                sqlalchemy.text(sql),
+                parameters={"table": table},
+            )
+            .mappings()
+            .all()
+        )
+
+    return [result["COLUMN_NAME"] for result in results]
+
+
+def _get_temp_table(table: str, run_id: str) -> TempTable:
+    """
+    Create a temp. table for data ingest.
+
+    Data will be stored in a temp. table before data are copied into
+    the target table. Here, we create the temp. table and appropriate
+    metadata.
+
+    Note: the `SELECT…INTO…` pattern assures that the temp. table will
+    be structurally identical to the target table.
+
+    :param table: target table to which data will be moved
+    :param run_id: unique identifier for processing
+    :return: temp. table details
+    """
+    # TODO: replace when #1884 merged.
+    temp_table_name = f"{table}_{run_id}_temp".replace("-", "_")
+
+    logger.info(f"Creating temp. table: {temp_table_name}.")
+    sql = textwrap.dedent(f"""
+    SELECT *
```

Comment:
Generally, these changes lean towards deriving as much from the DB as possible: here, for instance, the temp. table is derived from the target table, not defined in the code here.